// mattermost-redux/reducers/requests/groups.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _redux = require("redux");

var _action_types = require("../../action_types");

var _helpers = require("./helpers");

// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
// @flow
function linkGroupSyncable()
/*: RequestStatusType*/
{
  var state
  /*: RequestStatusType*/
  = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : (0, _helpers.initialRequestState)();
  var action
  /*: GenericAction*/
  = arguments.length > 1 ? arguments[1] : undefined;
  return (0, _helpers.handleRequest)(_action_types.GroupTypes.LINK_GROUP_SYNCABLE_REQUEST, _action_types.GroupTypes.LINK_GROUP_SYNCABLE_SUCCESS, _action_types.GroupTypes.LINK_GROUP_SYNCABLE_FAILURE, state, action);
}

function unlinkGroupSyncable()
/*: RequestStatusType*/
{
  var state
  /*: RequestStatusType*/
  = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : (0, _helpers.initialRequestState)();
  var action
  /*: GenericAction*/
  = arguments.length > 1 ? arguments[1] : undefined;
  return (0, _helpers.handleRequest)(_action_types.GroupTypes.UNLINK_GROUP_SYNCABLE_REQUEST, _action_types.GroupTypes.UNLINK_GROUP_SYNCABLE_SUCCESS, _action_types.GroupTypes.UNLINK_GROUP_SYNCABLE_FAILURE, state, action);
}

function getGroupSyncables()
/*: RequestStatusType*/
{
  var state
  /*: RequestStatusType*/
  = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : (0, _helpers.initialRequestState)();
  var action
  /*: GenericAction*/
  = arguments.length > 1 ? arguments[1] : undefined;
  return (0, _helpers.handleRequest)(_action_types.GroupTypes.GET_GROUP_SYNCABLES_REQUEST, _action_types.GroupTypes.GET_GROUP_SYNCABLES_SUCCESS, _action_types.GroupTypes.GET_GROUP_SYNCABLES_FAILURE, state, action);
}

function getGroupMembers()
/*: RequestStatusType*/
{
  var state
  /*: RequestStatusType*/
  = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : (0, _helpers.initialRequestState)();
  var action
  /*: GenericAction*/
  = arguments.length > 1 ? arguments[1] : undefined;
  return (0, _helpers.handleRequest)(_action_types.GroupTypes.GET_GROUP_MEMBERS_REQUEST, _action_types.GroupTypes.GET_GROUP_MEMBERS_SUCCESS, _action_types.GroupTypes.GET_GROUP_MEMBERS_FAILURE, state, action);
}

function getGroup()
/*: RequestStatusType*/
{
  var state
  /*: RequestStatusType*/
  = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : (0, _helpers.initialRequestState)();
  var action
  /*: GenericAction*/
  = arguments.length > 1 ? arguments[1] : undefined;
  return (0, _helpers.handleRequest)(_action_types.GroupTypes.GET_GROUP_REQUEST, _action_types.GroupTypes.GET_GROUP_SUCCESS, _action_types.GroupTypes.GET_GROUP_FAILURE, state, action);
}

function getAllGroupsAssociatedToTeam()
/*: RequestStatusType*/
{
  var state
  /*: RequestStatusType*/
  = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : (0, _helpers.initialRequestState)();
  var action
  /*: GenericAction*/
  = arguments.length > 1 ? arguments[1] : undefined;
  return (0, _helpers.handleRequest)(_action_types.GroupTypes.GET_ALL_GROUPS_ASSOCIATED_TO_TEAM_REQUEST, _action_types.GroupTypes.GET_ALL_GROUPS_ASSOCIATED_TO_TEAM_SUCCESS, _action_types.GroupTypes.GET_ALL_GROUPS_ASSOCIATED_TO_TEAM_FAILURE, state, action);
}

function getAllGroupsAssociatedToChannel()
/*: RequestStatusType*/
{
  var state
  /*: RequestStatusType*/
  = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : (0, _helpers.initialRequestState)();
  var action
  /*: GenericAction*/
  = arguments.length > 1 ? arguments[1] : undefined;
  return (0, _helpers.handleRequest)(_action_types.GroupTypes.GET_ALL_GROUPS_ASSOCIATED_TO_CHANNEL_REQUEST, _action_types.GroupTypes.GET_ALL_GROUPS_ASSOCIATED_TO_CHANNEL_SUCCESS, _action_types.GroupTypes.GET_ALL_GROUPS_ASSOCIATED_TO_CHANNEL_FAILURE, state, action);
}

function getGroupsAssociatedToTeam()
/*: RequestStatusType*/
{
  var state
  /*: RequestStatusType*/
  = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : (0, _helpers.initialRequestState)();
  var action
  /*: GenericAction*/
  = arguments.length > 1 ? arguments[1] : undefined;
  return (0, _helpers.handleRequest)(_action_types.GroupTypes.GET_GROUPS_ASSOCIATED_TO_TEAM_REQUEST, _action_types.GroupTypes.GET_GROUPS_ASSOCIATED_TO_TEAM_SUCCESS, _action_types.GroupTypes.GET_GROUPS_ASSOCIATED_TO_TEAM_FAILURE, state, action);
}

function getGroupsAssociatedToChannel()
/*: RequestStatusType*/
{
  var state
  /*: RequestStatusType*/
  = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : (0, _helpers.initialRequestState)();
  var action
  /*: GenericAction*/
  = arguments.length > 1 ? arguments[1] : undefined;
  return (0, _helpers.handleRequest)(_action_types.GroupTypes.GET_GROUPS_ASSOCIATED_TO_CHANNEL_REQUEST, _action_types.GroupTypes.GET_GROUPS_ASSOCIATED_TO_CHANNEL_SUCCESS, _action_types.GroupTypes.GET_GROUPS_ASSOCIATED_TO_CHANNEL_FAILURE, state, action);
}

function getGroupsNotAssociatedToTeam()
/*: RequestStatusType*/
{
  var state
  /*: RequestStatusType*/
  = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : (0, _helpers.initialRequestState)();
  var action
  /*: GenericAction*/
  = arguments.length > 1 ? arguments[1] : undefined;
  return (0, _helpers.handleRequest)(_action_types.GroupTypes.GET_GROUPS_NOT_ASSOCIATED_TO_TEAM_REQUEST, _action_types.GroupTypes.GET_GROUPS_NOT_ASSOCIATED_TO_TEAM_SUCCESS, _action_types.GroupTypes.GET_GROUPS_NOT_ASSOCIATED_TO_TEAM_FAILURE, state, action);
}

function getGroupsNotAssociatedToChannel()
/*: RequestStatusType*/
{
  var state
  /*: RequestStatusType*/
  = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : (0, _helpers.initialRequestState)();
  var action
  /*: GenericAction*/
  = arguments.length > 1 ? arguments[1] : undefined;
  return (0, _helpers.handleRequest)(_action_types.GroupTypes.GET_GROUPS_NOT_ASSOCIATED_TO_CHANNEL_REQUEST, _action_types.GroupTypes.GET_GROUPS_NOT_ASSOCIATED_TO_CHANNEL_SUCCESS, _action_types.GroupTypes.GET_GROUPS_NOT_ASSOCIATED_TO_CHANNEL_FAILURE, state, action);
}

var _default = ((0, _redux.combineReducers)({
  linkGroupSyncable: linkGroupSyncable,
  unlinkGroupSyncable: unlinkGroupSyncable,
  getGroupSyncables: getGroupSyncables,
  getGroupMembers: getGroupMembers,
  getGroup: getGroup,
  getAllGroupsAssociatedToTeam: getAllGroupsAssociatedToTeam,
  getAllGroupsAssociatedToChannel: getAllGroupsAssociatedToChannel,
  getGroupsAssociatedToTeam: getGroupsAssociatedToTeam,
  getGroupsAssociatedToChannel: getGroupsAssociatedToChannel,
  getGroupsNotAssociatedToTeam: getGroupsNotAssociatedToTeam,
  getGroupsNotAssociatedToChannel: getGroupsNotAssociatedToChannel
})
/*: (GroupsRequestsStatuses, GenericAction) => GroupsRequestsStatuses*/
);

exports.default = _default;
