// @hackmd/idle-js/dist/idle.js
'use strict';

function noop() {
}
// Idle.js main class
class Idle {
    /**
     * Initialize the class
     * @param {IdleOption} options
     */
    constructor(options) {
        this.isAway = false;
        // set default timeout to 3 seconds
        this.awayTimeout = 3000;
        this.awayTimestamp = 0;
        this.awayTimer = null;
        // events for monitoring user activity on the page
        this.onAway = noop;
        this.onAwayBack = noop;
        // events for the visibility API
        this.onVisible = noop;
        this.onHidden = noop;
        this.listener = null;
        if (options) {
            this.awayTimeout = parseInt(options.awayTimeout, 10);
            this.onAway = options.onAway || noop;
            this.onAwayBack = options.onAwayBack || noop;
            this.onVisible = options.onVisible || noop;
            this.onHidden = options.onHidden || noop;
        }
        const activeMethod = () => {
            return this.onActive();
        };
        // the methods that we will use to know when there is some activity on the page
        window.addEventListener('click', activeMethod);
        window.addEventListener('mousemove', activeMethod);
        window.addEventListener('mouseenter', activeMethod);
        window.addEventListener('keydown', activeMethod);
        window.addEventListener('scroll', activeMethod);
        window.addEventListener('mousewheel', activeMethod);
        window.addEventListener('touchmove', activeMethod);
        window.addEventListener('touchstart', activeMethod);
    }
    onActive() {
        this.awayTimestamp = new Date().getTime() + this.awayTimeout;
        if (this.isAway) {
            this.onAwayBack();
            this.start();
        }
        this.isAway = false;
        return true;
    }
    start() {
        if (!this.listener) {
            this.listener = (events) => {
                this.handleVisibilityChange();
            };
            document.addEventListener("visibilitychange", this.listener, false);
            document.addEventListener("webkitvisibilitychange", this.listener, false);
            document.addEventListener("msvisibilitychange", this.listener, false);
        }
        this.awayTimestamp = new Date().getTime() + this.awayTimeout;
        if (this.awayTimer !== null) {
            clearTimeout(this.awayTimer);
        }
        this.awayTimer = window.setTimeout(() => {
            this.checkAway();
        }, this.awayTimeout + 100);
        return this;
    }
    stop() {
        if (this.awayTimer != null) {
            clearTimeout(this.awayTimer);
        }
        if (this.listener !== null) {
            document.removeEventListener("visibilitychange", this.listener);
            document.removeEventListener("webkitvisibilitychange", this.listener);
            document.removeEventListener("msvisibilitychange", this.listener);
            this.listener = null;
        }
        return this;
    }
    setAwayTimeout(ms) {
        this.awayTimeout = ms;
        return this;
    }
    checkAway() {
        const t = new Date().getTime();
        if (t < this.awayTimestamp) {
            this.isAway = false;
            this.awayTimer = window.setTimeout(() => {
                this.checkAway();
            }, this.awayTimestamp - t + 100);
            return;
        }
        // away now
        if (this.awayTimer != null) {
            clearTimeout(this.awayTimer);
        }
        this.isAway = true;
        this.onAway();
    }
    handleVisibilityChange() {
        // check for hidden for various browsers
        if (document.hidden || document.msHidden || document.webkitHidden) {
            this.onHidden();
        }
        else {
            this.onVisible();
        }
    }
}

module.exports = Idle;
