// mattermost-redux/reducers/entities/groups.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("core-js/modules/es7.object.get-own-property-descriptors");

require("core-js/modules/es6.array.iterator");

require("core-js/modules/es6.object.keys");

require("core-js/modules/es7.symbol.async-iterator");

require("core-js/modules/es6.symbol");

require("core-js/modules/web.dom.iterable");

require("core-js/modules/es6.array.find-index");

var _redux = require("redux");

var _action_types = require("../../action_types");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { keys.push.apply(keys, Object.getOwnPropertySymbols(object)); } if (enumerableOnly) keys = keys.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function syncables() {
  var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var action = arguments.length > 1 ? arguments[1] : undefined;

  switch (action.type) {
    case _action_types.GroupTypes.RECEIVED_GROUP_TEAMS:
      {
        return _objectSpread({}, state, _defineProperty({}, action.group_id, _objectSpread({}, state[action.group_id], {
          teams: action.data
        })));
      }

    case _action_types.GroupTypes.RECEIVED_GROUP_CHANNELS:
      {
        return _objectSpread({}, state, _defineProperty({}, action.group_id, _objectSpread({}, state[action.group_id], {
          channels: action.data
        })));
      }

    case _action_types.GroupTypes.LINKED_GROUP_TEAM:
      {
        var nextGroupTeams = [];

        if (!state[action.data.group_id] || !state[action.data.group_id].teams) {
          nextGroupTeams = [action.data];
        } else {
          nextGroupTeams = _objectSpread({}, state)[action.data.group_id].teams;

          for (var i = 0, len = nextGroupTeams.length; i < len; i++) {
            if (nextGroupTeams[i].team_id === action.data.team_id) {
              nextGroupTeams[i] = action.data;
            }
          }
        }

        return _objectSpread({}, state, _defineProperty({}, action.data.group_id, _objectSpread({}, state[action.data.group_id], {
          teams: nextGroupTeams
        })));
      }

    case _action_types.GroupTypes.LINKED_GROUP_CHANNEL:
      {
        var nextGroupChannels = [];

        if (!state[action.data.group_id] || !state[action.data.group_id].channels) {
          nextGroupChannels = [action.data];
        } else {
          nextGroupChannels = _objectSpread({}, state)[action.data.group_id].channels;

          for (var _i = 0, _len = nextGroupChannels.length; _i < _len; _i++) {
            if (nextGroupChannels[_i].channel_id === action.data.channel_id) {
              nextGroupChannels[_i] = action.data;
            }
          }
        }

        return _objectSpread({}, state, _defineProperty({}, action.data.group_id, _objectSpread({}, state[action.data.group_id], {
          channels: nextGroupChannels
        })));
      }

    case _action_types.GroupTypes.UNLINKED_GROUP_TEAM:
      {
        if (!state[action.data.group_id]) {
          return state;
        }

        var nextTeams = state[action.data.group_id].teams.slice();
        var index = nextTeams.findIndex(function (groupTeam) {
          return groupTeam.team_id === action.data.syncable_id;
        });

        if (index !== -1) {
          nextTeams.splice(index, 1);
        }

        return _objectSpread({}, state, _defineProperty({}, action.data.group_id, _objectSpread({}, state[action.data.group_id], {
          teams: nextTeams
        })));
      }

    case _action_types.GroupTypes.UNLINKED_GROUP_CHANNEL:
      {
        if (!state[action.data.group_id]) {
          return state;
        }

        var nextChannels = state[action.data.group_id].channels.slice();

        var _index = nextChannels.findIndex(function (groupChannel) {
          return groupChannel.channel_id === action.data.syncable_id;
        });

        if (_index !== -1) {
          nextChannels.splice(_index, 1);
        }

        return _objectSpread({}, state, _defineProperty({}, action.data.group_id, _objectSpread({}, state[action.data.group_id], {
          channels: nextChannels
        })));
      }

    default:
      return state;
  }
}

function members() {
  var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var action = arguments.length > 1 ? arguments[1] : undefined;

  switch (action.type) {
    case _action_types.GroupTypes.RECEIVED_GROUP_MEMBERS:
      {
        return _objectSpread({}, state, _defineProperty({}, action.group_id, {
          members: action.data.members,
          totalMemberCount: action.data.total_member_count
        }));
      }

    default:
      return state;
  }
}

function groups() {
  var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var action = arguments.length > 1 ? arguments[1] : undefined;

  switch (action.type) {
    case _action_types.GroupTypes.RECEIVED_GROUP:
      {
        return _objectSpread({}, state, _defineProperty({}, action.data.id, action.data));
      }

    case _action_types.GroupTypes.RECEIVED_GROUPS:
      {
        var nextState = _objectSpread({}, state);

        var _iteratorNormalCompletion = true;
        var _didIteratorError = false;
        var _iteratorError = undefined;

        try {
          for (var _iterator = action.data[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
            var group = _step.value;
            nextState[group.id] = group;
          }
        } catch (err) {
          _didIteratorError = true;
          _iteratorError = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion && _iterator.return != null) {
              _iterator.return();
            }
          } finally {
            if (_didIteratorError) {
              throw _iteratorError;
            }
          }
        }

        return nextState;
      }

    default:
      return state;
  }
}

var _default = (0, _redux.combineReducers)({
  syncables: syncables,
  members: members,
  groups: groups
});

exports.default = _default;
